engine/execution/ingestion/engine.go

```
@@ -930,344 +992,422 @@ func (e *Engine) generateExecutionReceipt(
+		Logger()
 
+	// ignore requests for unsealed height
+	if req.FromHeight > sealedHeight {
+		log.Info().Msg("receives state sync requests for unsealed height, ignore")
 		return nil
-	})
+	}
+
+	// fromHeight, toHeight must be sealed height
+	fromHeight, toHeight := req.FromHeight, req.ToHeight
+	if toHeight > sealedHeight {
+		toHeight = sealedHeight
+	}
+
+	// for each height starting from fromHeight to toHeight,
+	// query the statecommitment, and if exists, send the
+	// state delta
+	// TOOD: add context
+	ctx := e.unit.Ctx()
+	err = e.deltaRange(ctx, fromHeight, toHeight,
+		func(delta *messages.ExecutionStateDelta) {
+			err := e.syncConduit.Submit(delta, originID)
+			if err != nil {
+				e.log.Error().Err(err).Msg("could not submit block delta")
+			}
+		})
+
+	if err != nil {
+		return fmt.Errorf("could not send deltas: %w", err)
+	}
+
+	log.Info().Msg("responded state deltas for a height range")
+
+	return nil
 }
 
-func (e *Engine) saveDelta(ctx context.Context, executionStateDelta *messages.ExecutionStateDelta) {
+// deltaRange querys it's local execution state to find deltas for a height
+// range between the fromHeight to the toHeight. If delta is found, then
+// pass it to the onDelta callback.
+func (e *Engine) deltaRange(ctx context.Context, fromHeight uint64, toHeight uint64,
+	onDelta func(*messages.ExecutionStateDelta)) error {
+
+	for height := fromHeight; height <= toHeight; height++ {
+		header, err := e.state.AtHeight(height).Head()
+		if err != nil {
+			return fmt.Errorf("could not query block header at height: %v", height)
+		}
 
-	defer e.syncWg.Done()
+		blockID := header.ID()
+		_, err = e.execState.StateCommitmentByBlockID(ctx, blockID)
 
+		if err == nil {
+			// this block has been executed, we will send the delta
+			delta, err := e.execState.RetrieveStateDelta(ctx, blockID)
+			if err != nil {
+				return fmt.Errorf("could not retrieve state delta for block %v, %w", blockID, err)
+			}
+
+			onDelta(delta)
```

Should we ignore error here?

---

The error is already handled [here](https://github.com/onflow/flow-go/pull/31/files#diff-9d11a07967b04bd9f51ebd6b01a170d5R1200-R1202)